### ISSUE TITLE
Travis builds the same commit twice (once in a branch and once in PR)

### ISSUE THREAD

**Issue by denik** [travis-ci/travis-ci#1147]
I understand that when testing PR travis actually checks how it would be when merged to master and when testing a branch it just pulls that branch. This is good.

However, in some cases the branch is already rebased against master, so no merge needed and travis ends up building the same commit twice.

We have that happening a lot in our travis pro account. Would be nice if travis could be smarter about this and only build the same commit once.

**Comment by marutosi** [comment]
I think this was Travis behavior in some years ago.
I don't know when Travis changed behavior to build the same commit twice.

**Comment by marutosi** [comment]
> I think this was Travis behavior in some years ago.

It means Travis builds the same commit only once in some years ago.

**Comment by marutosi** [comment]
> It means Travis builds the same commit only once in some years ago.

There is same comment.
https://github.com/travis-ci/travis-ci/issues/3276#issue-58798010

**Comment by joshk** [comment]
Travis has always build the branch and PR, as they are essentially two different things, one being a development branch, and the other being the merge commit from GitHub.

It is only recently that we have exposed this information using the commit status API as GitHub made a change allowing for multiple commit status updates.

There are ways to disable branch builds but allow for PR builds to continue.

I'm going to close this issue as we are no looking at changing the current default behaviour.

Thanks for all the comments and feedback on this issue.

**Comment by iphydf** [comment]
> Travis has always build the branch and PR, as they are essentially two different things, one being a development branch, and the other being the merge commit from GitHub.
This is not what the issue is about. It's about travis building the *same* commit twice, even when fast-forwarded. Building a merge commit as separate build makes sense, but building the exact same hash twice in different branches does not.

**Comment by joshk** [comment]
Sorry @iphydf, I am a bit confused. This specific issue is about Travis building the commit for the branch and then the PR.

We only have access to information about the merge commit, even then it is limited. For that reason we suggest devs who want us to only test the PR should then disable builds for branches where the development is happening. We will still test the PR as long as the branch it is being merged into is being tested to.

**Comment by marutosi** [comment]
I don't know why @joshk close this issue.
This issue suggests if PR is fast-forwared-able and travis has building result on **forked** repository,
travis should not re-run building.
It is easy to know whether fast-forwared-able or not.

**Comment by sethladd** [comment]
Can you elaborate on this: "There are ways to disable branch builds but allow for PR builds to continue."

How do we only build for master and Pull Requests ? I'd like to disable builds for other branches, as they always have a Pull Request for them.

**Comment by Ehekatl** [comment]
+1

**Comment by marutosi** [comment]
Why not repopen this issue?

**Comment by whit537** [comment]
> Can you elaborate on this: "There are ways to disable branch builds but allow for PR builds to continue."
>
How do we only build for master and Pull Requests ? I'd like to disable builds for other branches, as they always have a Pull Request for them.

@sethladd [Whitelist](https://docs.travis-ci.com/user/customizing-the-build/#Building-Specific-Branches) `master` by adding this to your `.travis.yml`:

```yml
branches:
  only:
    - "master"
```

The merge commit will still trigger a build, because it's a merge into `master` (assuming your PR is against `master`, of course :wink:). As long as the branch whence the PR originates is *not* whitelisted, the commit on that branch *won't* trigger a build.

**Comment by nijikokun** [comment]
The above line breaks tagged commit deploys however in many cases.

**Comment by rpavlovs** [comment]
This issue should be reopened.

**Comment by MiguelMadero** [comment]
I think we can detect that the commit was built or there is a build in progress already and avoid doing another commit for the same one.

**Comment by portnov** [comment]
Hello.
People are doing

```
branches:
  only:
    master
```

But, wouldn't this disable builds triggering at the event of push into PR branch after PR was created?

**Comment by ryanhiebert** [comment]
> But, wouldn't this disable builds triggering at the event of push into PR branch after PR was created?

No, pull requests against master also build with this configuration, including updates to the Pull Request by pushing more commits to that branch.

**Comment by portnov** [comment]
Thanks for the information.

**Comment by temoto** [comment]
Is it possible to solve this problem "from client"? Something like:

```
# travis.yml

pre-install:
  curl "travis/api/get-build-by-commit?project=foo&commit=$TRAVIS_COMMIT"
  if build.finished then skip
```

**Comment by jonhoo** [comment]
FYI, in [`rust-imap`](https://github.com/jonhoo/rust-imap) I use [this rule](https://github.com/jonhoo/rust-imap/blob/30079d8b9bc22fe64d893849262908101381e002/.travis.yml#L10), which works pretty well:
```yaml
if: type != push OR branch = master OR branch =~ /^v\d+\.\d+(\.\d+)?(-\S*)?$/
```
Specifically, it will test any commit to master, any version tag, and anything that is not a push (so things like PRs).

**Comment by junaruga** [comment]
I would just share another way to prevent Travis from building the same commit twice "still enabling Travis at push on the forked repository".

Here is the commit using the `if` syntax.
https://github.com/ruby/ruby/commit/f7741cda323e759432f34c954d7da0ac3d865ab1

Here is the related ticket on Travis community.
https://travis-ci.community/t/11673
